python/ray/autoscaler/v2/scheduler.py

```diff
@@ -0,0 +1,346 @@
+
+
+@dataclass
+class ResourceScheduleRequest:
+    # TODO: This prob could be refactored into the ClusterStatus data class later.
+    # The current ray resource requests.
+    resource_requests: List[ResourceRequestByCount] = field(default_factory=list)
+    # The Gang resource requests.
+    gang_resource_requests: List[GangResourceRequest] = field(default_factory=list)
+    # cluster resource constraints.
+    cluster_resource_constraints: List[ClusterResourceConstraint] = field(
+        default_factory=list
+    )
+    # The ray nodes
+    node_states: List[NodeState] = field(default_factory=list)
+    # The current list of instances.
+    current_instances: List[Instance] = field(default_factory=list)
+    # The resource schedule config.
+    schedule_config: ResourceScheduleConfig
+
+
+@dataclass
+class ResourceScheduleReply:
+    # The infeasible resource bundles.
+    infeasible_resource_requests: List[ResourceRequestByCount] = field(
+        default_factory=list
+    )
+    # The infeasible gang resource bundles.
+    infeasible_gang_resource_requests: List[GangResourceRequest] = field(
+        default_factory=list
+    )
+    # The infeasible cluster resource constraints.
+    infeasible_cluster_resource_constraints: List[ClusterResourceConstraint] = field(
+        default_factory=list
+    )
+    # The target cluster shape, given the current resource demands and instances.
+    # Key is the node type name, value is the number of nodes.
+    # Note this might be "smaller" than the current cluster shape, since there
+    # could be resource requests constraints enforced.
+    target_cluster_shape: Dict[NodeType, int]
+    # The current cluster shape.
+    current_cluster_shape: Dict[NodeType, int]
+
+
+class IResourceScheduler(metaclass=ABCMeta):
+    """
+    Interface for a resource scheduler.
+
+    Implements the `instance_manager.proto ResourceSchedulerService` interface.
+    """
+
+    @abstractmethod
+    def schedule_resource_bundles(
+        self, request: ResourceScheduleRequest
+    ) -> ResourceScheduleReply:
+        """
+        Given the resource requests and the current cluster state, calculate the
+        target cluster shape by trying to schedule the resource requests on the
+        nodes.
+        """
+        pass
+
+
+class SchedulingNodeStatus(Enum):
+    """
+    The status of a scheduling node (`SchedulingNode`)
+    """
+
+    # The node is to be launched.
+    TO_LAUNCH = "TO_LAUNCH"
+    # The node is pending to be launched (a pending request, or a pending instance)
+    PENDING = "PENDING"
+    # The node is running.
+    RUNNING = "RUNNING"
+
+
+@dataclass
+class SchedulingNode:
+    """
+    A abstraction of a node that can be scheduled on by the resource scheduler.
+
+    A scheduling node is expected to be used as:
+
+        node  = SchedulingNode.from_node_config(node_config)
+        infeasible = node.try_schedule(requests)
+        score = node.compute_score()
```

Comment:
should we combine these two methods into one. Seems they will always be called together.